examples/knative/go/kn-go-tagging/tagging/vsphere.go

```diff
@@ -0,0 +1,176 @@
+package tagging
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"time"
+
+	"github.com/kelseyhightower/envconfig"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/session"
+	"github.com/vmware/govmomi/session/keepalive"
+	"github.com/vmware/govmomi/vapi/rest"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/methods"
+	"github.com/vmware/govmomi/vim25/soap"
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	// DefaultMountPath is where the vSphere secret will be saved.
+	DefaultMountPath = "/var/bindings/vsphere"
+	// vCenter APIs keep-alive time interval.
+	keepAliveInterval = 5 * time.Minute
+)
+
+// ReadKey reads the key from the secret.
+func ReadKey(key string) (string, error) {
+	var env EnvConfig
+	if err := envconfig.Process("", &env); err != nil {
+		return "", err
+	}
+
+	mountPath := DefaultMountPath
+	if env.SecretPath != "" {
+		mountPath = env.SecretPath
+	}
+
+	data, err := ioutil.ReadFile(filepath.Join(mountPath, key))
+	if err != nil {
+		return "", err
+	}
+	return string(data), nil
+}
+
+// newSOAPClient returns a vCenter SOAP API client with active keep-alive. Use
+// Logout() to release resources and perform a clean logout from vCenter.
+func newSOAPClient(ctx context.Context) (*govmomi.Client, error) {
+	var env EnvConfig
+	if err := envconfig.Process("", &env); err != nil {
+		return nil, err
+	}
+
+	parsedURL, err := soap.ParseURL(env.VCAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read the username and password from the filesystem.
+	username, err := ReadKey(corev1.BasicAuthUsernameKey)
+	if err != nil {
+		return nil, err
+	}
+	password, err := ReadKey(corev1.BasicAuthPasswordKey)
+	if err != nil {
+		return nil, err
+	}
+	parsedURL.User = url.UserPassword(username, password)
+
+	soapClient := soap.NewClient(parsedURL, env.Insecure)
+	vimClient, err := vim25.NewClient(ctx, soapClient)
+	if err != nil {
+		return nil, fmt.Errorf("new VIM client: %w", err)
+	}
+
+	vSphereClient := govmomi.Client{
+		Client: vimClient,
+	}
+
+	return &vSphereClient, nil
+}
+
+func newSOAPClientWithKeepAlive(ctx context.Context, soapURL *url.URL) (*govmomi.Client, error) {
```

Comment:
We don't need both; we can just use the soap client with keep alive and call it "newSOAPClient". It will be consistent with newRESTClient(...) which has keep alive but doesn't have "keepAlive" in the function name.